File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/sliding_window/P438_FindAllAnagramsInAString.py
from typing import List
from collections import defaultdict


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        n, m = len(s), len(p)
        if n < m:
            return []
        need, window = defaultdict(int), defaultdict(int)
        for c in p:
            need[c] += 1
        left, right, valid = 0, 0, 0
        ans = []
        while right < n:
            # expand window
            c = s[right]
            right += 1
            if c in need:
                window[c] += 1
                if need[c] == window[c]:
                    valid += 1
            else:
                window.clear()
                left = right
                valid = 0
            # shrink window
            if right - left == m:
                if valid == len(need):
                    ans.append(left)
                d = s[left]
                left += 1
                if need[d] == window[d]:
                    valid -= 1
                window[d] -= 1
        return ans


if __name__ == '__main__':
    s = "cbaebabacd"
    p = "abc"
    solu = Solution()
    print(solu.findAnagrams(s, p))

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/array/sliding_window/P438_FindAllAnagramsInaString.java
package com.dsa.leetcode.labuladong.array.sliding_window;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @since 2023/11/25 10:36
 */
public class P438_FindAllAnagramsInaString {
    public List<Integer> findAnagrams(String s, String p) {
        int n = s.length(), m = p.length();
        if (n < m) {
            return new ArrayList<>();
        }
        List<Integer> ans = new ArrayList<>();

        Map<Character, Integer> need = new HashMap<>();
        for (char c : p.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }

        Map<Character, Integer> window = new HashMap<>();
        int left = 0, right = 0;
        int valid = 0;

        while (right < n) {
            // expand window
            char c = s.charAt(right);
            right++;
            if (need.containsKey(c)) {
                window.put(c, window.getOrDefault(c, 0) + 1);
                if (need.get(c).equals(window.get(c))) {
                    valid++;
                }
            } else {
                window.clear();
                valid = 0;
                left = right;
            }

            // shrink window
            if (right - left == m) {
                if (valid == need.size()) {
                    ans.add(left);
                }
                char d = s.charAt(left);
                left++;
                if (need.get(d).equals(window.get(d))) {
                    valid--;
                }
                window.put(d, window.get(d) - 1);
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        String s = "abaacbabc", p = "abc";
        P438_FindAllAnagramsInaString find = new P438_FindAllAnagramsInaString();
        System.out.println(find.findAnagrams(s, p));
    }
}
